File: src/integrationTest/java/net/woggioni/gbcs/test/AbstractServerTest.java
package net.woggioni.gbcs.test;


import net.woggioni.gbcs.GradleBuildCacheServer;
import net.woggioni.gbcs.api.Configuration;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestMethodOrder;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.io.TempDir;

import java.nio.file.Path;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
public abstract class AbstractServerTest {

    protected Configuration cfg;
    protected Path testDir;
    private GradleBuildCacheServer.ServerHandle serverHandle;

    @BeforeAll
    public void setUp0(@TempDir Path tmpDir) {
        this.testDir = tmpDir;
        setUp();
        startServer(cfg);
    }

    @AfterAll
    public void tearDown0() {
        tearDown();
        stopServer();
    }

    protected abstract void setUp();

    protected abstract void tearDown();

    private void startServer(Configuration cfg) {
        this.serverHandle = new GradleBuildCacheServer(cfg).run();
    }

    private void stopServer() {
        if (serverHandle != null) {
            try (GradleBuildCacheServer.ServerHandle handle = serverHandle) {
                handle.shutdown();
            }
        }
    }
}

File: src/test/kotlin/net/woggioni/gbcs/test/AbstractServerTest.kt
package net.woggioni.gbcs.test

import net.woggioni.gbcs.GradleBuildCacheServer
import net.woggioni.gbcs.api.Configuration
import org.junit.jupiter.api.AfterAll
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestMethodOrder
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path


@TestInstance(TestInstance.Lifecycle.PER_CLASS)
@TestMethodOrder(MethodOrderer.OrderAnnotation::class)
abstract class AbstractServerTest {

    protected lateinit var cfg : Configuration

    protected lateinit var testDir : Path

    private var serverHandle : GradleBuildCacheServer.ServerHandle? = null

    @BeforeAll
    fun setUp0(@TempDir tmpDir : Path) {
        this.testDir = tmpDir
        setUp()
        startServer(cfg)
    }

    @AfterAll
    fun tearDown0() {
        tearDown()
        stopServer()
    }

    abstract fun setUp()

    abstract fun tearDown()

    private fun startServer(cfg : Configuration) {
        this.serverHandle = GradleBuildCacheServer(cfg).run()
    }

    private fun stopServer() {
        this.serverHandle?.use {
            it.shutdown()
        }
    }
}